rasa/core/policies/ensemble.py

```diff
@@ -578,16 +578,29 @@ def _pick_best_policy(
 
         form_confidence = None
         form_policy_name = None
-        # End-to-end predictions overrule all other predictions.
-        use_only_end_to_end = any(
+        # different type of predictions have different priorities
+        # No user predictions overrule all other predictions.
+        is_no_user_prediction = any(
+            prediction.is_no_user_prediction for prediction in predictions.values()
+        )
+        # End-to-end predictions overrule all other predictions based on user input.
+        is_end_to_end_prediction = any(
             prediction.is_end_to_end_prediction for prediction in predictions.values()
         )
-        policy_events = []
 
+        policy_events = []
         for policy_name, prediction in predictions.items():
             policy_events += prediction.events
 
-            if prediction.is_end_to_end_prediction != use_only_end_to_end:
+            # No user predictions overrule all other predictions.
+            if prediction.is_no_user_prediction != is_no_user_prediction:
+                continue
+
+            # End-to-end predictions overrule all other predictions based on user input.
+            if (
+                not is_no_user_prediction
+                and prediction.is_end_to_end_prediction != is_end_to_end_prediction
```

Comment:
we don't have to check for `not is_no_user_prediction` again, do we?
```suggestion
                if is_end_to_end_prediction and not prediction.is_end_to_end_prediction:
```

Reply:
we do. It's so happen that right now, `is_end_to_end_prediction` will be `False`. It happens because in reality we need three values rather than 2 booleans, but since you wanted 2 booleans, I'd keep the check here

Reply:
I still think having `None` would be significantly more clear implementation